File: 04/CSharp/QueueApp/AbstractQueue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


namespace QueueApp
{
    public abstract class AbstractQueue : IQueue
    {
        abstract public void Enqueue(Object element);

        abstract public Object Element();

        abstract public Object Dequeue();

        abstract public void Clear();

        abstract public int Size();

        abstract public bool IsEmpty();
        
        /*
        abstract public IQueue MakeCopy();
        
        public IQueue Filter(Predicate<Object> predicate)
        {
            IQueue ret = MakeCopy();
            int s = ret.Size();

            for (int i = 0; i < s; i++)
            {
                Object elem = ret.Dequeue();
                //if (predicate.target(elem))
                {
                    ret.Enqueue(elem);
                }
            }

            return ret;
        }

        public IQueue Map(Func<Object, Object> func)
        {
            IQueue ret = MakeCopy();
            int s = ret.Size();
            for (int i = 0; i < s; i++)
            {
                ret.Enqueue(ret.Dequeue());
            }

            return ret;
        }
        */
    }
}

File: 04/Java/AbstractQueue.java
import java.util.function.Predicate;
import java.util.function.Function;

abstract public class AbstractQueue implements Queue {

    abstract public void enqueue(Object element);

    abstract public Object element();
    
    abstract public Object dequeue();
    
    abstract public void clear();

    abstract public int size();

    abstract public boolean isEmpty();

    abstract public AbstractQueue makeCopy();

    public AbstractQueue filter(Predicate<Object> predicate) {
        AbstractQueue ret = makeCopy();
        int s = ret.size();

        for (int i = 0; i < s; i++) {
            Object elem = ret.dequeue();
            if (predicate.test(elem)) {
                ret.enqueue(elem);
            }
        }

        return ret;
    }

    public AbstractQueue map(Function<Object, Object> func) {
        AbstractQueue ret = makeCopy();
        int s = ret.size();
        for (int i = 0; i < s; i++) {
            ret.enqueue(func.apply(ret.dequeue()));
        }

        return ret;
    }
}